composer/datasets/streaming/writer.py

```diff
@@ -0,0 +1,150 @@
+
+    def _flush_shard(self) -> None:
+        """Flush cached samples to a new dataset shard."""
+        os.makedirs(self.dirname, exist_ok=True)
+        shard = len(self.samples_per_shard)
+        basename = get_shard_basename(shard)
+        filename = os.path.join(self.dirname, basename)
+        with open(filename, 'xb') as out:
+            for data in self.new_samples:
+                out.write(data)
+        self.samples_per_shard.append(len(self.new_samples))
+        self.bytes_per_shard.append(self.new_shard_size)
+        self.new_samples = []
+        self.new_shard_size = 0
+
+    def _write_index(self) -> None:
+        """Save dataset index file."""
+        if self.new_samples:
+            raise RuntimeError("Attempted to write index file while samples are still being processed.")
+        filename = os.path.join(self.dirname, get_index_basename())
+        ndarray_samples_per_shard = np.asarray(self.samples_per_shard, np.int64)
+        ndarray_bytes_per_shard = np.asarray(self.bytes_per_shard, np.int64)
+        ndarray_bytes_per_sample = np.asarray(self.bytes_per_sample, np.int64)
+        index = StreamingDatasetIndex(ndarray_samples_per_shard, ndarray_bytes_per_shard, ndarray_bytes_per_sample,
+                                      self.fields)
+        with open(filename, 'xb') as out:
+            index.dump(out)
+
+    def write_sample(self, sample: Dict[str, bytes]) -> None:
+        """Add a sample to the dataset.
+
+        Args:
+            sample (Dict[str, bytes]): The new sample, whose keys must contain the fields to save (others ignored).
+        """
+        data = sample_dict_to_bytes(sample, self.fields)
+        if self.shard_size_limit <= self.new_shard_size + len(data):
+            self._flush_shard()
+        self.bytes_per_sample.append(len(data))
+        self.new_samples.append(data)
+        self.new_shard_size += len(data)
+
+    def write_samples(self,
+                      samples: Iterable[Dict[str, bytes]],
+                      use_tqdm: bool = True,
+                      total: Optional[int] = None) -> None:
+        """Add the samples from the given iterable to the dataset.
+
+        Args:
+            samples (Iterable[Dict[str, bytes]]): The new samples.
+            use_tqdm (bool): Whether to display a progress bar.  Default: ``True``.
+            total (int, optional): Total samples for the progress bar (for when samples is a generator).
+        """
+        if use_tqdm:
+            samples = tqdm(samples, leave=False, total=total)
+        for s in samples:
+            self.write_sample(s)
+
+    def finish(self) -> None:
+        """Complete writing the dataset by flushing last samples to a last shard, then write an index file."""
+        if self.new_samples:
+            self._flush_shard()
+        self._write_index()
+
+    def __enter__(self):
+        """Enter as contextmanager."""
+        return self
+
+    def __exit__(self, *args: List, **kwargs: Dict) -> None:
```

Comment:
`from types import TracebackType`
`from typing import Type, Optional`

```suggestion
    def __exit__(self, exc_type: Optional[Type[BaseException]], exc: Optional[BaseException], traceback: Optional[TracebackType]) -> None:
```